From d301cbb992e80218735b81c73d49fec705f91d08 Mon Sep 17 00:00:00 2001
From: Bjarne von Horn <bjarne.von.horn@rwth-aachen.de>
Date: Sun, 1 May 2022 16:39:10 +0200
Subject: [PATCH] Fix setting mac_address

eoe->dev->dev_addr is read_only in 5.17.0
using eth_hw_addr_set instead
---
 devices/ccat/netdev.c | 10 ++++++++++
 devices/generic.c     |  4 ++++
 master/debug.c        |  4 ++++
 3 files changed, 18 insertions(+), 0 deletions(-)

diff --git a/devices/ccat/netdev.c b/devices/ccat/netdev.c
index 5ff1159e..44d83607 100644
--- a/devices/ccat/netdev.c
+++ b/devices/ccat/netdev.c
@@ -894,8 +894,18 @@ static int ccat_eth_init_netdev(struct ccat_eth_priv *priv)
 	int status;
 
 	/* init netdev with MAC and stack callbacks */
+
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 15, 0)
+	u8 mac_addr[ETH_ALEN];
+
+	if (priv->netdev->addr_len != ETH_ALEN)
+		return -EFAULT;
+	memcpy_fromio(mac_addr, priv->reg.mii + 8, ETH_ALEN);
+	eth_hw_addr_set(priv->netdev, mac_addr);
+#else
 	memcpy_fromio(priv->netdev->dev_addr, priv->reg.mii + 8,
 		      priv->netdev->addr_len);
+#endif
 	priv->netdev->netdev_ops = &ccat_eth_netdev_ops;
 
 	/* use as EtherCAT device? */
diff --git a/devices/generic.c b/devices/generic.c
index 0feab284..4d4d9bbd 100644
--- a/devices/generic.c
+++ b/devices/generic.c
@@ -255,7 +255,11 @@ int ec_gen_device_offer(
     int ret = 0;
 
     dev->used_netdev = desc->netdev;
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 15, 0)
+    eth_hw_addr_set(dev->netdev, desc->dev_addr);
+#else
     memcpy(dev->netdev->dev_addr, desc->dev_addr, ETH_ALEN);
+#endif
 
     dev->ecdev = ecdev_offer(dev->netdev, ec_gen_poll, THIS_MODULE);
     if (dev->ecdev) {
diff --git a/master/debug.c b/master/debug.c
index ad5f90eb..b4bfc442 100644
--- a/master/debug.c
+++ b/master/debug.c
@@ -134,7 +134,11 @@ void ec_debug_register(
     ec_debug_unregister(dbg);
 
     // use the Ethernet address of the physical device for the debug device
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 15, 0)
+    eth_hw_addr_set(dbg->dev, net_dev->dev_addr);
+#else
     memcpy(dbg->dev->dev_addr, net_dev->dev_addr, ETH_ALEN);
+#endif
 
     // connect the net_device to the kernel
     if ((result = register_netdev(dbg->dev))) {
-- 
GitLab

